tokio-codec/src/lines_codec.rs

```diff
@@ -38,22 +107,72 @@ fn without_carriage_return(s: &[u8]) -> &[u8] {
 
 impl Decoder for LinesCodec {
     type Item = String;
+    // TODO: in the next breaking change, this should be changed to a custom
+    // error type that indicates the "max length exceeded" condition better.
     type Error = io::Error;
 
     fn decode(&mut self, buf: &mut BytesMut) -> Result<Option<String>, io::Error> {
-        if let Some(newline_offset) =
-            buf[self.next_index..].iter().position(|b| *b == b'\n')
-        {
-            let newline_index = newline_offset + self.next_index;
-            let line = buf.split_to(newline_index + 1);
-            let line = &line[..line.len()-1];
-            let line = without_carriage_return(line);
-            let line = utf8(line)?;
-            self.next_index = 0;
-            Ok(Some(line.to_string()))
-        } else {
-            self.next_index = buf.len();
-            Ok(None)
+        loop {
+            // Determine how far into the buffer we'll search for a newline. If
+            // there's no max_length set, we'll read to the end of the buffer.
+            let read_to = if let Some(limit) = self.max_length {
+                // If a max length is set, read max_length bytes from the
+                // *current* offset into the buffer (which may not be 0).
+                let read_to = limit + self.next_index + 1;
+                // Clamp to buf.len(), so we don't go out of bounds.
+                cmp::min(read_to, buf.len())
+            } else {
+                buf.len()
+            };
+
+            let newline_offset = buf[self.next_index..read_to]
+                .iter()
+                .position(|b| *b == b'\n');
+
+            if self.is_discarding {
+                let discard_to = if let Some(offset) = newline_offset {
```

Comment:
Maybe the contents of this `if` clause could be moved to a `fn discard` to help simplify this function.

Reply:
@carllerche I made that change in 9b2cda0f0cd3d652cfcfceced2bd862c7c3f46af. This actually ended up being the only one of your suggestions that I'm not totally sure about --- I'm not sure if moving this to its own function actually makes the code clearer or if it just makes the logic harder to follow. What do you think?